src/stream.cpp

```diff
@@ -69,22 +69,24 @@ namespace oxen::quic
 
     void Stream::close(uint64_t error_code)
     {
-        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
-
-        if (is_shutdown)
-            log::info(log_cat, "Stream is already shutting down");
-        else if (is_closing)
-            log::debug(log_cat, "Stream is already closing");
-        else
-        {
-            is_closing = is_shutdown = true;
-            log::info(log_cat, "Closing stream (ID: {}) with error code {}", stream_id, ngtcp2_strerror(error_code));
-            ngtcp2_conn_shutdown_stream(conn, stream_id, error_code);
-        }
-        if (is_shutdown)
-            data_callback = {};
+        conn.quic_manager->call([this, error_code]() {
+            log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
```

Comment:
small nitpick, but can you move these trace logging function names to be between the async call and the function header? thinking in case it doesn't print the correct function name since its inside a callable

```
void Stream::close(uint64_t error_code)
{
        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
        conn.quic...
```

Reply:
`__PRETTY_FUNCTION__` is evaluated by the pre-processor at compile time which has no idea that the string is being put in a function object to be called later, so it will have the correct function name.  I put it inside the lambda so it prints when the meat of the function executes rather than when it's queued, as logs could get confusing then.

Reply:
__PRETTY_FUNCTION__ is going to be ugly because it's in a lambda, but should still have the outer function name in it.  So still functional, but ugly.

Reply:
Maybe you want *both* traces?  i.e. so you trace when it's called, and then trace when it's actually called inside the event loop thread.

Reply:
Changed to use `source_location` and slightly custom `oxen::logging` invocation